### PR TITLE
* added fingerprint authentication, added extensions ;)

### PR DESCRIPTION
**Pull request by SerhiiOst**

### CODE REVIEW

File: app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+) : FingerprintManagerCompat.AuthenticationCallback() {
+
+    private val manager = FingerprintManagerCompat.from(context)
+
+    private var keyStore: KeyStore? = null
+    private val keyName = "ONSEO"
+    private var cipher: Cipher? = null
+
+    //TODO clean that code
+    fun authenticate() {
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt]: Simplify this code, to mach nested **if**`s.

File: app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            keyGenerator!!.init(
+                KeyGenParameterSpec.Builder(
+                    keyName,
+                    KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
+                ).setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setUserAuthenticationRequired(true)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7).build()
+            )
+            keyGenerator.generateKey()
+        } catch (e: IOException) {
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt]: Simplify this, you don`t handle all errors, you just show it to console, so you can make just one exception

File: app/src/main/java/ua/onseo/ds/ui/login/FragmentLogin.kt [hunk 3]
```diff
@@ -22,7 +29,12 @@ class FragmentLogin : Fragment() {
...
     override fun onActivityCreated(savedInstanceState: Bundle?) {
         super.onActivityCreated(savedInstanceState)
-        viewModel = ViewModelProviders.of(this).get(LoginViewModel::class.java)
+        viewModel = getViewModel(LoginViewModel::class.java)
+        handleFingerprint()
         // TODO: Use the ViewModel
     }
+
+    private fun handleFingerprint() {
+        FingerprintHandler(context!!).authenticate()
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/login/FragmentLogin.kt]: avoid call !!, try to use let

File: app/src/main/java/ua/onseo/ds/utils/BiometricUtils.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import android.content.pm.PackageManager
+import android.Manifest.permission.USE_BIOMETRIC
+import android.Manifest.permission.USE_FINGERPRINT
+import android.content.Context
+import android.support.v4.app.ActivityCompat
+import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
+import android.os.Build
+import android.support.annotation.RequiresApi
+
+//YES, I COPY-PASTED ALL OF THIS FROM MEDIUM DON'T JUDGE ME!!!!!
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/utils/BiometricUtils.kt]: Nice try man, nice try=)

**Comment by YPoudanenOnSeo** [comment]
@SerhiiOst resolve all conflicts to continue

File: app/src/main/java/ua/onseo/ds/ui/home_screen/FragmentHome.kt [hunk 5]
```diff
@@ -35,7 +30,21 @@ class FragmentHome : FragmentBase() {
...
+        askForFingerprintLogin()
+
+    }
+
+    private fun askForFingerprintLogin() {
+        if (!viewModel.isFirstFingerprint())
+            return
+
+        AlertDialog.Builder(context)
+            .setTitle("Fingerprint login")
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/home_screen/FragmentHome.kt]: Move cardcoded strings to strings.xml

File: app/src/main/java/ua/onseo/ds/ui/home_screen/FragmentHome.kt [hunk 6]
```diff
@@ -35,7 +30,21 @@ class FragmentHome : FragmentBase() {
...
+            .setMessage("Do you want to enable fingerprint login?")
+            .setPositiveButton("Yes") { _, _ -> viewModel.setFingerprintLogin(true) }
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/home_screen/FragmentHome.kt]: Move cardcoded strings to strings.xml

File: app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            keyGenerator.generateKey()
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+
+    @TargetApi(Build.VERSION_CODES.M)
+    private fun cipherInit(): Boolean =
+        try {
+            cipher =
+                Cipher.getInstance(KeyProperties.KEY_ALGORITHM_AES + "/" + KeyProperties.BLOCK_MODE_CBC + "/" + KeyProperties.ENCRYPTION_PADDING_PKCS7)
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt]: Use Kotlin way to concat strings.

File: app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+class FingerprintHandler(
+    private val context: Context,
+    private val loginFunction: () -> Unit,
+    private val errorFunction: () -> Unit
+) : FingerprintManagerCompat.AuthenticationCallback() {
+
+    private val manager = FingerprintManagerCompat.from(context)
+
+    private var keyStore: KeyStore? = null
+    private val keyName = context.getString(R.string.FINGERPRINT_KEY)
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/login/FingerprintHandler.kt]: Don`t use strings key with upper case.

File: app/src/main/java/ua/onseo/ds/ui/completed_list/FragmentCompletedList.kt [hunk 9]
```diff
@@ -2,16 +2,29 @@ package ua.onseo.ds.ui.completed_list
...
+import android.view.ViewGroup
 import ua.onseo.ds.R
+import ua.onseo.ds.utils.getViewModel
 import ua.onseo.ds.ui.FragmentBase
 
 class FragmentCompletedList : FragmentBase() {
 
     private lateinit var viewModel: CompletedListViewModel
 
+    override fun onCreateView(
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/ui/completed_list/FragmentCompletedList.kt]: remove this, you must to implement method from FragmentBase

File: app/src/main/java/ua/onseo/ds/repository/api/SharedPrefsHelper.kt [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        val tokenExpiredTimeMillis = this.sharedPreferences?.getLong(this.PREFS_TOKEN_EXPIRED_TIME, 0)
+        return (tokenExpiredTimeMillis != null && (System.currentTimeMillis() < tokenExpiredTimeMillis))
+    }
+
+    fun askForFingerprintLogin(): Boolean {
+        return sharedPreferences!!.getBoolean(PREFS_FINGERPRINT_LOGIN_ASK, true)
+    }
+
+    fun setAskForFingerprintLogin(value: Boolean) {
+        sharedPreferences!!.edit().also {
```
Comment (YPoudanenOnSeo) [inline app/src/main/java/ua/onseo/ds/repository/api/SharedPrefsHelper.kt]: Try to avoid use !!

**Comment by SerhiiOst** [comment]
please, just accept this pull request finally))

**Pull request merged**